### ISSUE TITLE
Bogus "no suitable conversion" error in Intellisense on Mac/clang

### ISSUE THREAD

**Issue by zajo** [microsoft/vscode-cpptools#1395]
I got this error from Intellisense:

On this source line:

```
shared_ptr<connection_list> cl=tlsd.cl_.lock();
```

Intellisense says:

```
no suitable user-defined conversion from
"boost::shared_ptr<boost::synapse::synapse_detail::thread_local_signal_data::connection_list>" to
"boost::shared_ptr<boost::synapse::synapse_detail::thread_local_signal_data::connection_list>" exists
```
The compiler itself doesn't issue an error. Somehow Intellisense succeeded deducing the same correct fully qualified name for both the type of `cl` and the return type of `tlsd.cl_.lock()`, but then it erroneously decided they differ.

VSCode version 1.19.1 (1.19.1), 0759f77bb8d86658bc935a10a64f6182c5a1eeba, C/C++ extension version 0.14.5, Apple LLVM version 9.0.0 (clang-900.0.39.2).

Thanks!

**Comment by sean-mcmanus** [comment]
In order to fix this, we need an isolated repro. I have boost_1_66_0, but I don't see any synapse class. How do we get that? Or if you can create a simplified repro with a dummy class.

**Comment by zajo** [comment]
Synapse is not an official Boost library. I can't easily give you a small repro case at this time, unfortunately.

**Comment by sean-mcmanus** [comment]
I messed around with various shared_ptr<customClass>, but I couldn't repro the issue. If you want us to fix, this you'd need to binary search on the definition of the connection_list to identify the C++ feature that is causing this. It could be some C++17 feature, since full support hasn't been added yet.

**Comment by zajo** [comment]
I think I have something to help you diagnose this: in this library (http://zajo.github.io/boost-synapse/), there is a configuration macro, `BOOST_SYNAPSE_USE_STD_SMART_PTR`, which can be used to control whether it uses `boost::shared_ptr` or `std::shared_ptr`, and it never refers to shared_ptr with its fully qualified name. When I added this macro to `c_cpp_properties.json`, the bogus intellisense error disappeared. It's still strange because in the intellisense error it _did_ show both types as `boost::shared_ptr`.

**Comment by sean-mcmanus** [comment]
Might be the same as https://github.com/Microsoft/vscode-cpptools/issues/1559 . It was marked as fixed at https://developercommunity.visualstudio.com/content/problem/198343/incorrect-c-intellisense-error-no-suitable-user-de.html, but it turns out it still repros...so no ETA on an actual fix.

UPDATE: https://developercommunity.visualstudio.com/content/problem/286802/vs2017-pro-editor-complains-polymorphism-smart-poi.html .

**Comment by sean-mcmanus** [comment]
The bug you experienced should be fixed with 0.18.0, although there is a new bug with the same error that started as of 0.17.7 (see https://github.com/Microsoft/vscode-cpptools/issues/2422 ).
